metrics/internal/factory/factory.go

```diff
@@ -0,0 +1,68 @@
+package factory
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/mozilla-services/rapid-release-model/metrics/internal/config"
+	"github.com/mozilla-services/rapid-release-model/metrics/internal/export"
+	"github.com/mozilla-services/rapid-release-model/metrics/internal/github"
+	"github.com/shurcooL/githubv4"
+	"golang.org/x/oauth2"
+)
+
+// Factory is used to create dependencies for the CLI application
+type Factory struct {
+	NewEncoder             func() (export.Encoder, error)
+	NewExporter            func() (export.Exporter, error)
+	NewGitHubHTTPClient    func() (*http.Client, error)
+	NewGitHubGraphQLClient func() (github.GraphQLClient, error)
+	NewGitHubRepo          func() (*github.Repo, error)
+}
+
+// NewFactory creates the default Factory for the CLI application
+func NewFactory(ctx context.Context) *Factory {
+	f := new(Factory)
+
+	f.NewEncoder = func() (export.Encoder, error) {
+		return &export.JSONEcoder{}, nil
+	}
+
+	f.NewExporter = func() (export.Exporter, error) {
+		encoder, err := f.NewEncoder()
+		if err != nil {
+			return nil, err
+		}
+		return export.NewWriterExporter(os.Stdout, encoder)
+	}
+
+	f.NewGitHubHTTPClient = func() (*http.Client, error) {
+		token := config.FromEnv("GITHUB", "TOKEN")
+		if token == "" {
+			return nil, fmt.Errorf("Requires GitHub token to be set in env")
```

Comment:
This will be covered by the upcoming docs and feels nitpicky, but it would be nice to print the full envvar we're looking for here too: `RRM_METRICS__GITHUB__TOKEN`

Reply:
That's good feedback, @dlactin! I just pushed a commit to improve error reporting for this. Thanks!